### PR TITLE
Simpler MasterOnly MPI mechanism

### PR DESCRIPTION
**Pull request by SimonHeybrock**
Replaces dummy workspaces by NULL/None workspaces. This seems to make handling `MasterOnly` execution and implementation of such algorithms easier.

There is a minor drawback to this approach: Previously Python scripts could rely on obtained a default-constructed (but uninitialized) workspace on non-master ranks. This allow for, e.g.,

```python
ws = MyAlg()
for i in range ws.getNumberHistograms():
  pass
```
`getNumberHistograms()` returns 0 if the worksapce is not initialized so this piece of code worked. Now `ws` is `None`, so we have to write:

```python
ws = MyAlg()
if ws is not None:
  for i in range ws.getNumberHistograms():
    pass
```
This is slightly more verbose and might necessitate slightly more changes in some reduction scripts. On the other hand it may also catch some actual issues -- code that is actually broken for a parallel workflow -- so this change could maybe also be seen as having a positive effect.

**To test:**

Code review.
No issue, no release notes, since MPI is still internal.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/API/test/AlgorithmMPITest.h
Comment (OwenArnold) [inline Framework/API/test/AlgorithmMPITest.h]: Looks reasonable to me.

**Review by OwenArnold — APPROVED** [review]
Changes look reasonable to me. I would suggest a second review from another TSC member before merging.

**Comment by SimonHeybrock** [comment]
I have encountered another potential drawback of this approach: Some code may base decisions on the existance of a workspaces (example: SANS2D reused reduced can run, result is only on master rank, non-master ranks thus re-reduce). Any more complex check on a workspace would fail, so unless the check is *purely* based on the workspace name this drawback does not apply.

I will give this some more thought based on those thoughts.

**Comment by SimonHeybrock** [comment]
Regarding my previous comment: I believe testing for existing workspaces in the ADS is a corner case (which should probably be discouraged anyway) and we should not let this influence the overall design, i.e., this should go ahead as is. Removing the in-progress flag.

**Review by martyngigg — APPROVED** [review]
I think this looks sensible. I like that we're able to remove the boilerplate `execNonMaster` implementations - that feels cleaner.

I agree that a non-master rank receiving "null" workspace is a better indication that the previous uninitialized value.

**Pull request merged**
